File: app/src/main/java/com/my/test2/BaseAdapter.java
package com.my.test2;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewbinding.ViewBinding;

/**
 * Created by Administrator on 16-7-18.
 */
public abstract class BaseAdapter<T, B extends ViewBinding> extends RecyclerView.Adapter<BaseAdapter.ViewHolder> {
    protected OnItemClick onItemClick;
    protected List<T> mDatas;
    protected Context context;

    public List<T> getmDatas() {
        return mDatas;
    }

    public void setmDatas(List<T> mDatas) {
        this.mDatas = mDatas;
    }

    public BaseAdapter(Context context) {
        this.context = context;
    }

    public BaseAdapter(Context context, List<T> mDatas) {
        this.mDatas = mDatas;
        this.context = context;
    }

    public OnItemClick getOnItemClick() {
        return onItemClick;
    }

    public void setOnItemClick(OnItemClick onItemClick) {
        this.onItemClick = onItemClick;

    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(context);
        B binding = setBinding(inflater, parent);
        ViewHolder<B> viewHolder = new ViewHolder<B>(binding.getRoot());
        viewHolder.setBinding(binding);
        if(onItemClick!=null){
            viewHolder.setOnItemClick(onItemClick);
        }
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        ViewHolder<B> myholder=(ViewHolder<B>) holder;
        onBindHolder(myholder,position);
    }
    protected abstract void onBindHolder(ViewHolder<B> holder, int position);

    @Override
    public int getItemViewType(int position) {
        return position;
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    protected abstract B setBinding(LayoutInflater inflater, ViewGroup parent);

    @Override
    public int getItemCount() {
        return mDatas == null ? 0 : mDatas.size();
    }

    public static class ViewHolder<B extends ViewBinding> extends RecyclerView.ViewHolder implements View.OnClickListener {
        protected OnItemClick onItemClick;
        public B binding;

        public OnItemClick getOnItemClick() {
            return onItemClick;
        }

        public void setOnItemClick(OnItemClick onItemClick) {
            this.onItemClick = onItemClick;
        }

        public B getBinding() {
            return binding;
        }

        public void setBinding(B binding) {
            this.binding = binding;
        }

        public ViewHolder(View itemView) {
            super(itemView);
        }
        public void setClick(View... views){
            if(onItemClick!=null) {
                for (View view : views) {
                    view.setOnClickListener(this);
                }
            }
        }

        @Override
        public void onClick(View v) {
            onItemClick.onClick(v,getAdapterPosition());
        }
    }

    public static interface  OnItemClick{
        void onClick(View view, int position);
    }

}

File: app/src/main/java/com/my/test2/kotlinbase/BaseAdapter.kt
package com.my.test2.kotlinbase
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding

/**
 * Created by Administrator on 16-7-18.
 */
abstract class BaseAdapter<T, B : ViewBinding?> :
    RecyclerView.Adapter<BaseAdapter.ViewHolder<B>> {
    var onItemClick:OnItemClick? = null
    protected var mDatas: kotlin.collections.List<T>? = null
    protected var context: android.content.Context
    fun getmDatas(): kotlin.collections.List<T>? {
        return mDatas
    }

    fun setmDatas(mDatas:List<T>?){
        this.mDatas = mDatas
    }

    constructor(context: android.content.Context) {
        this.context = context
    }

    constructor(context: android.content.Context, mDatas: List<T>?) {
        this.mDatas = mDatas
        this.context = context
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): BaseAdapter.ViewHolder<B> {
        val inflater = LayoutInflater.from(context)
        val binding = setBinding(inflater, parent)
        val viewHolder =
            BaseAdapter.ViewHolder<B>(binding!!.root)
        viewHolder.setBindingInstance(binding)
        if (onItemClick != null) {
            viewHolder.onItemClick = onItemClick
        }
        return viewHolder
    }

    override fun onBindViewHolder(
        holder: ViewHolder<B>,
        position: Int
    ) {
        val myholder =
            holder
        onBindHolder(myholder, position)
    }

    protected abstract fun onBindHolder(
        holder: BaseAdapter.ViewHolder<B>?,
        position: Int
    )

    override fun getItemViewType(position: Int): Int {
        return position
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    protected abstract fun setBinding(inflater: LayoutInflater?, parent: ViewGroup?): B
    override fun getItemCount(): Int {
        return if (mDatas == null) 0 else mDatas!!.size
    }

    class ViewHolder<B : ViewBinding?>(itemView: android.view.View) :
        androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView), android.view.View.OnClickListener {
        var onItemClick: BaseAdapter.OnItemClick? = null
        var binding: B? = null

        fun setBindingInstance(binding: B): kotlin.Unit {
            this.binding = binding
        }

        fun setClick(vararg views: android.view.View){
            if (onItemClick != null) {
                for (view in views) {
                    view.setOnClickListener(this)
                }
            }
        }

        override fun onClick(v: android.view.View) {
            onItemClick?.onClick(v, getAdapterPosition())
        }
    }

    interface OnItemClick {
        fun onClick(view: android.view.View?,position: Int);
    }
}